kanalony-web-server/app/controllers/Application.scala

```diff
@@ -0,0 +1,126 @@
+import argonaut._
+import Argonaut._
+import scala.concurrent._
+
+class Application extends Controller {
+
+  def query: Action[AnyContent] = Action.async({ implicit request => doQuery(request) })
+
+  def doQuery(request : Request[AnyContent]) : Future[Result] = {
+    try {
+      val analyticsRequestOption = for {
+        input <- request.body.asJson
+        req <- input.toString.decodeOption[AnalyticsRequest]
+      } yield req
+
+      if (analyticsRequestOption.isEmpty) {
+        return wrapWithPromise(BadRequest("Invalid input - ensure proper JSON is supplied with the mandatory fields supplied and that the supplied MIME type is application/json"))
+      }
+
+      val queryParams = requestToQueryParams(analyticsRequestOption.get)
+      val queryExecutionResult = execute(queryParams)
+      val queryResponseData =  queryExecutionResult map { data => AnalyticsResponse(data.headers, data.rows) }
+
+      queryResponseData map {
+        data => Ok(data.asJson.toString)
+      } recover {
+        case e : Exception => BadRequest(e.getMessage)
+      }
+    }
+    catch {
+      case e: QueryNotSupportedException => {
+        return wrapWithPromise(BadRequest(e.getMessage))
+      }
+      case e: IllegalArgumentException => {
+        return wrapWithPromise(BadRequest(e.getMessage))
+      }
+      case e: Exception => {
+        println(e.getClass.getTypeName)
+        return wrapWithPromise(InternalServerError(e.getMessage))
+      }
+    }
+  }
+
+  def wrapWithPromise(res : Result): Future[Result] = {
+    return (Promise[Result] success res).future
+  }
+
+  def execute(queryParams: QueryParams): Future[IQueryResult] = {
+    QueryLocator.locate(queryParams).query(queryParams)
+  }
+
+  def extractValues[T, E <: Exception](values : List[String], enumConverter : String => T, exceptionCreator : String => E): List[T] = {
+    var processedValue = ""
+    var res = List[T]()
+    try {
+      values.foreach(d => {
+        processedValue = d
+        res = res :+ enumConverter(d)
+      })
+      res
+    }
+    catch {
+      case e : Exception => throw exceptionCreator(processedValue)
+    }
+  }
+
+  def extractMetrics(metrics : List[String]): List[Metrics.Value] = {
+    extractValues(metrics, Metrics.withName(_), s => new InvalidMetricsException("Metric " + s + " not supported"))
+  }
+
+  def extractDimension(dimension : String): Dimensions.Value = {
+    val results = extractValues(List(dimension), Dimensions.withName(_), s => new InvalidDimensionException(s))
+    results(0)
+  }
+
+  def extractDimensions(dimensions : List[String]): List[Dimensions.Value] = {
+    extractValues(dimensions, Dimensions.withName(_), s => new InvalidDimensionException(s))
+  }
+
+  def createConstraint(dimension: Dimensions.Value, values: List[String]) : IDimensionConstraint = {
+    try {
+      dimension match {
+        case Dimensions.partner => new DimensionEqualityConstraint[Int](values.toSet.map { (v: String) => v.toInt })
```

Comment:
I wanted to discuss with you the following:
What if we treat partner like regular filtered dimension? it will be provided by the wrapping PHP which authenticates the request, to allow regular equality constraint.
What do you think?